touch/touch.cpp

```diff
@@ -0,0 +1,439 @@
+			x = (mouse_x + (event->tfinger.dx * 1.25 * speedFactor * DISPLAY_WIDTH));
+			y = (mouse_y + (event->tfinger.dy * 1.25 * speedFactor * DISPLAY_HEIGHT));
+		}
+
+		x = CLIP(x, 0, (int)DISPLAY_WIDTH);
+		y = CLIP(y, 0, (int)DISPLAY_HEIGHT);
+		xrel = x - mouse_x;
+		yrel = y - mouse_y;
+
+		// update the current finger's coordinates so we can track it later
+		for (int i = 0; i < MAX_NUM_FINGERS; i++) {
+			if (finger[port][i].id != id)
+			continue;
+			finger[port][i].last_x = x;
+			finger[port][i].last_y = y;
+		}
+
+		// If we are starting a multi-finger drag, start holding down the mouse button
+		if (num_fingers_down >= 2 && !multi_finger_dragging[port]) {
+			// only start a multi-finger drag if at least two fingers have been down long enough
+			int num_fingers_downlong = 0;
+			for (int i = 0; i < MAX_NUM_FINGERS; i++) {
+				if (finger[port][i].id == NO_TOUCH) {
+					continue;
+				}
+				if (event->tfinger.timestamp - finger[port][i].time_last_down > MAX_TAP_TIME) {
+					num_fingers_downlong++;
+				}
+			}
+			if (num_fingers_downlong >= 2) {
+				int mouse_down_x = mouse_x;
+				int mouse_down_y = mouse_y;
+				if (direct_touch) {
+					for (int i = 0; i < MAX_NUM_FINGERS; i++) {
+						if (finger[port][i].id == id) {
+							uint32_t earliest_time = finger[port][i].time_last_down;
+							for (int j = 0; j < MAX_NUM_FINGERS; j++) {
+								if (finger[port][j].id >= 0 && (i != j) ) {
+									if (finger[port][j].time_last_down < earliest_time) {
+										mouse_down_x = finger[port][j].last_x;
+										mouse_down_y = finger[port][j].last_y;
+										earliest_time = finger[port][j].time_last_down;
+									}
+								}
+							}
+							break;
+						}
+					}
+				}
+
+				Uint8 simulated_button = 0;
+				if (num_fingers_downlong == 2) {
+					simulated_button = SDL_BUTTON_LEFT;
+					multi_finger_dragging[port] = DRAG_TWO_FINGER;
+				} else {
+					simulated_button = SDL_BUTTON_RIGHT;
+					multi_finger_dragging[port] = DRAG_THREE_FINGER;
+				}
+				SDL_Event ev;
+				set_mouse_button_event(&ev, SDL_MOUSEBUTTONDOWN, simulated_button, mouse_down_x, mouse_down_y);
+				SDL_PushEvent(&ev);
+			}
+		}
+
+		if (!xrel && !yrel) {
+			return;
+		}
+
+		// check if this is the "oldest" finger down (or the only finger down)
+		// otherwise it will not affect mouse motion
+		bool update_pointer = true;
+		if (num_fingers_down > 1) {
+			for (int i = 0; i < MAX_NUM_FINGERS; i++) {
+				if (finger[port][i].id != id) {
+					continue;
+				}
+				for (int j = 0; j < MAX_NUM_FINGERS; j++) {
+					if (finger[port][j].id == NO_TOUCH || (j == i)) {
+						continue;
+					}
+					if (finger[port][j].time_last_down < finger[port][i].time_last_down) {
+						update_pointer = false;
+					}
+				}
+			}
+		}
+		if (!update_pointer) {
+			return;
+		}
+		set_mouse_motion_event(event, x, y, xrel, yrel);
+	}
+}
+
+void finish_simulated_mouse_clicks(int current_mouse_x, int current_mouse_y)
```

Comment:
We can just use `OutputToLogical` here, it's similar to this function but doesn't depend on hardcoded output size.

Reply:
This is finishing all the accumulated mouse clicks at the correct time. I doubt there's another function that does this, because it depends on other code in touch.cpp

Reply:
My comment somehow got relocated (I guess the branch was forced pushed-to while I was writing it). I posted this about `convert_touch_xy_to_game_xy`.

Reply:
Ah sorry. I think there's no point in writing comments if they move around and many force pushes happen :)

I also noticed when I click on "View Changes" github fails to locate the commits, probably also because of force pushes. Maybe I should wait with comments until the dust has settled...

Reply:
Ok, things should bee good now, feel free to coment the code :)